### PR TITLE
Technical Debt Culture Page

### PR DESCRIPTION
**Pull request by msaspence**
This PR will add a page on our culture for dealing with technical debt.

Note it intentionally omits the implementation for dealing with technical debt. Those details are less clear right now, and I wanted to limit the scope of the initial PR.

### CODE REVIEW

File: tech-debt.md [hunk 1]
```diff
@@ -2,5 +2,88 @@
 
 > a concept in programming that reflects the extra development work that arises when code that is easy to implement in the short run is used instead of applying the best overall solution
 
-We don't currently have a strategy for technical debt, at least not documented. This document aims to do that.
+Tech Debt is an inevitable part of programming, a valuable tool even. Do not shy away from it. Embrace it! Subdue it! Make it your servant. What ever you do, *do not just ignore it* and hope it goes away.
```
Comment (pmcnr-hx) [inline tech-debt.md]: what ever → whatever

File: tech-debt.md [hunk 2]
```diff
@@ -2,5 +2,88 @@
...
 
+*Note:* This document is aspirational, it doesn't necessarily reflect the current culture of technical debt at Holiday Extras. It does however reflect where we would like to be, and is a work in progress. If you have ideas about how to better manage technical debt please tell people. Good places to do this are the #developers, #guid-front-end and #guild-back-end slack channels or with a PR for this page.
```
Comment (pmcnr-hx) [inline tech-debt.md]: Correct channel names: **#guild-frontend-tech** and **#guild-backend-tech**

File: tech-debt.md (hunk 1 above)
Comment (msaspence) [inline tech-debt.md]: FYI you can edit typos like this yourself
- View full changes
- Pencil top left
- Make change
- Commit directly

Quicker that commenting and having the creator correct

File: tech-debt.md [hunk 3]
```diff
@@ -2,5 +2,88 @@
 
 > a concept in programming that reflects the extra development work that arises when code that is easy to implement in the short run is used instead of applying the best overall solution
 
-We don't currently have a strategy for technical debt, at least not documented. This document aims to do that.
+Tech Debt is an inevitable part of programming, a valuable tool even. Do not shy away from it. Embrace it! Subdue it! Make it your servant. Whatever you do, *do not just ignore it* and hope it goes away.
 
+*Note:* This document is aspirational, it doesn't necessarily reflect the current culture of technical debt at Holiday Extras. It does however reflect where we would like to be, and is a work in progress. If you have ideas about how to better manage technical debt please tell people. Good places to do this are the #developers, #guid-front-end and #guild-back-end slack channels or with a PR for this page.
```
Comment (pmcnr-hx) [inline tech-debt.md]: Correct channel names: **#guild-frontend-tech** and **#guild-backend-tech**

**Comment by hxpaul** [comment]
👍  some good points

File: tech-debt.md [hunk 4]
```diff
@@ -2,5 +2,88 @@
...
+
+These decisions have to consider the business demands and the technical implications. They must be made by pod leads and developers in concert. By their nature it is impossible for either to do it on their own.
+
+It is the responsibility of developers to explain to leads the current state of technical debt and any implications it will have on current or future work. It is the responsibility of leads to explain to developers the priorities of the business and the urgency and importance of work.
+
+When it comes to paying off existing debt, focus on debt that currently has or is likely to have in the near future a "high rate of interest". Technical debt that is isolated can be safely left for another day.
+
+### 6. Little and often
+
+Technical debt should be paid of continuously in small payments. Exactly how much is up for debate. But here are some things to bare in mind:
```
Comment (theninj4) [inline tech-debt.md]: *off

File: tech-debt.md [hunk 5]
```diff
@@ -2,5 +2,88 @@
...
+  * what, if anything, the changes do to help address existing tech debt
+  * what technical debt the changes introduce or compound, and any justification for them
+  * what GitHub issues and/or Jiras have been raised to address any new debt in the future
+* *Guilds* can be used to see if people in other pods are having a similar experience with an existing piece of tech debt.
+
+### 4. Limit its influence
+
+Technical debt that is isolated has much less scope to interfere with other code. This tends to emerge naturally if code is loosely coupled and modular in nature.
+
+Say you have a module and its internal implementation is mess, but its interface is well thought out and does not change very often. This is an example of isolated technical debt. The messy internals will rarely interfere with work on the wider application. If however the interface becomes the subject of many changes, or it has a lot of bugs, it might be a good time to refactor or even just throw it out and start again.
```
Comment (elliottcrush) [inline tech-debt.md]: Might be worth linking in here to the facade design pattern. It's a useful practice for isolating debt like this and might be a good place to give people a bit more information.

**Comment by michaelcarter** [comment]
I don't disagree with anything in here. 👍

File: tech-debt.md [hunk 6]
```diff
@@ -2,5 +2,88 @@
...
+  * they have some of the best knowledge on existing tech debt
+  * they will be able to advise how to avoid tech debt in the first place
+  * they can help make your case to others that a certain piece of tech debt needs to be addressed
+  * you can help keep them up to date with the general picture of tech debt
+
+#### Where to do this?
+
+The existing communication channels are perfect for talking about tech debt. We just need to make sure we are using them.
+
+* *Stand ups* can be a great place to flag any technical debt you are facing or expecting to face in the immediate term.
```
Comment (jamesmcneil) [inline tech-debt.md]: We could also suggest Scrum of Scrums, but I guess that is just a really long standup

File: tech-debt.md [hunk 7]
```diff
@@ -2,5 +2,88 @@
 
 > a concept in programming that reflects the extra development work that arises when code that is easy to implement in the short run is used instead of applying the best overall solution
 
-We don't currently have a strategy for technical debt, at least not documented. This document aims to do that.
+Tech Debt is an inevitable part of programming, a valuable tool even. Do not shy away from it. Embrace it! Subdue it! Make it your servant. Whatever you do, *do not just ignore it* and hope it goes away.
 
+*Note:* This document is aspirational, it doesn't necessarily reflect the current culture of technical debt at Holiday Extras. It does however reflect where we would like to be, and is a work in progress. If you have ideas about how to better manage technical debt please tell people. Good places to do this are the #developers, #guid-frontend-tech and #guild-backend-tech slack channels or with a PR for this page.
```
Comment (pmcnr-hx) [inline tech-debt.md]: Oops... still `#guid-frontend-tech` (`l` missing in `guild`)...

**Comment by theninj4** [comment]
I like what you've tried to do here, but I'm against a document like this until we can pin down what `tech debt` actually is. The current definition is too loose as it assumes the person making the judgement call is aware of `the best overall solution`, which will vary wildly between developers.

We need to draw a line in the sand that says to business owners "this is the line, everyone abides by the line, and we can't cross the line". Right now everyone's lines are in different places and push back rightly or wrongly in different situations under the banner of `tech debt` that muddies the term and makes it hard for us to use it when we really need to.

My take - encouraging more people to throw around the term is only going to make the situation worse.

File: tech-debt.md [hunk 8]
```diff
@@ -2,5 +2,88 @@
...
+This commandment comes from our general company values. All our values are relevant however this one particularly so.
+
+*Be courageous* and push back against pressure to ship more and more technical debt without considerations for the long term implications. Be respectful of other stakeholders but make sure that you speak up when you think the issue of technical debt is not being addressed correctly.
+
+### 2. Anticipate Tech Debt From the Beginning
+
+Tech Debt to some extent is part and parcel of every piece of code that you write. Before you lay down a single line of code, you should be thinking about tech debt. Perhaps even as soon as you are aware of upcoming work. It is never too early to be thinking about:
+
+* existing debt that might slow you down
+* existing debt that it makes sense to pay off
```
Comment (pmcnr-hx) [inline tech-debt.md]: remove the `it`. just `that makes sense`? (same below?)

File: tech-debt.md [hunk 9]
```diff
@@ -2,5 +2,88 @@
...
+  * they have some of the best knowledge on existing tech debt
+  * they will be able to advise how to avoid tech debt in the first place
+  * they can help make your case to others that a certain piece of tech debt needs to be addressed
+  * you can help keep them up to date with the general picture of tech debt
+
+#### Where to do this?
+
+The existing communication channels are perfect for talking about tech debt. We just need to make sure we are using them.
+
+* *Stand ups* can be a great place to flag any technical debt you are facing or expecting to face in the immediate term.
```
Comment (pmcnr-hx) [inline tech-debt.md]: `Stand-ups`?

File: tech-debt.md (hunk 4 above)
Comment (guymeurice) [inline tech-debt.md]: *bear in mind

Comment (pmcnr-hx) [inline tech-debt.md]: bare → bear

**Comment by pmcnr-hx** [comment]
Excellent work! 👍

File: tech-debt.md [hunk 10]
```diff
@@ -2,5 +2,88 @@
...
+
+* Little and often over big and rarely.
+* Big payoffs may be required from time to time.
+* Avoid big refactors that drag on for more than one sprint, aka "Refactor Purgatory".
+* If a big rewrite is required do everything you can to transition to a new design incrementally.
+* Even if a sprint solely addresses tech debt it should be split into iterative chunks if possible.
+
+### 7. Aspire to be a boy scout
+
+> Always leave the campground cleaner than you found it.
```
Comment (Nick-Patrick) [inline tech-debt.md]: is it possible to make this stand out more? i see it as one of the most important things we should focus on

**Comment by jackdclark** [comment]
👍 Yes

**Comment by msaspence** [comment]
@theninj4 I agree that we need a clearer definition of what tech debt is, that is what this document seeks to do, but I don't think drawing a line is helpful. "Saying we can't cross that line" is a recipe for being pushed across it again and again. The realities of the business mean life is more complex than this, there are grey areas and we need to better equip the team and individuals to handle these grey areas in a deliberate and considered way.

Since every piece of code we write has a shelf life, to an extent every piece of code is tech debt. Whether something is or isn't tech debt isn't the important part. The important questions to be asking are:
- what "interest" will we have to pay on this code in the short _and_ long term
- what "capital" will we have to pay on this code to prevent paying the expected interest

**Comment by elliottcrush** [comment]
👍

File: tech-debt.md [hunk 11]
```diff
@@ -2,5 +2,89 @@
...
+
+### 4. Limit its influence
+
+Technical debt that is isolated has much less scope to interfere with other code. This tends to emerge naturally if code is loosely coupled and modular in nature.
+
+Say you have a module and its internal implementation is mess, but its interface is well thought out and does not change very often. This is an example of isolated technical debt. The messy internals will rarely interfere with work on the wider application. If however the interface becomes the subject of many changes, or it has a lot of bugs, it might be a good time to refactor or even just throw it out and start again. [The Facade Pattern](https://en.wikipedia.org/wiki/Facade_pattern) is one way to do this.
+
+### 5. Be deliberate
+
+Whilst it is "okay" to take on technical debt, it has to be a deliberate decision. The short term gains of shipping it now have to be weighed against the long term interest that will need to be paid off over its lifetime, not to mention the debt itself.
```
Comment (hx-markterry) [inline tech-debt.md:66]: I think the team/pod/product owner needs to know what tech debt is being accepted so there are no surprises later when they need to address it or it slows them down later etc..

Comment (msaspence) [inline tech-debt.md:66]: @hx-markterry are you happy this is covered on line 70 now?

Comment (hx-markterry) [inline tech-debt.md:66]: yup happy now 😃

**Pull request merged**